### PR TITLE
fix: flaky test `Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx should queue send tx after switch network confirmation and transaction should target the correct network after switch is confirmed`

### LINKED ISSUE #27387: Fix "Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx should queue send tx after switch network confirmation and transaction should target the correct network after switch is confirmed" flaky test
### What is this about?

Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx should queue send tx after switch network confirmation and transaction should target the correct network after switch is confirmed

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by seaona**
## **Description**
There is a race condition with window management that makes this test fail quite often. The test performs these actions and there are different behaviors happening, depending on the build:

1. Action: trigger a network switch -> this opens a new dialog window (we have 4)
2. Action: trigger a send -> we are still with the network switch window open
3. Action: we click switch network -> this closes the dialog (now we have 3 windows for a brief moment - if this is fast, like in webpack, it fails with the error `Error: waitUntilXWindowHandles timed out polling window handles. Expected: 3, Actual: 4`)
4. A new dialog is open automatically -> now we have 4 windows again, but the dialog handle ids are different
5. Sometimes this dialog is closed and then a new one is open
6. We switch to the dialog, but often the context is invalidated as we are switching to either 1 or 4 or 5

Extra Notes:
This PR fixes the flakiness of waiting for 3 windows by adding a delay. This won't fix the root cause of the 2 new identified issues, but after re-runs has proven to stabilize the test

- In Webpack, sometimes the dialog for the transaction never appears, and the wallet get's in  a frozen state. The issue is on the wallet level, see https://github.com/MetaMask/metamask-extension/issues/27414
- see issue https://github.com/MetaMask/metamask-extension/issues/27360
Since there is only a brief moment for having 3 windows, waiting for 3 windows and then proceeding also makes the test flaky, so it's not effective.

Screenshot from 2024-09-26 11-26-34

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27352?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/27387

## **Manual testing steps**

1. Check ci
2. Run test locally `yarn test:e2e:single test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js --browser=chrome`

## **Screenshots/Recordings**

Problem in FF and Webpack, see how popup is open and closed, and then a new popup is open with the transaction.

https://github.com/user-attachments/assets/c17ff7bd-6b1d-4517-b33f-95f7ac6d3120

Problem in Webpack, where tx does not open after the change networks

https://github.com/user-attachments/assets/2b626159-ec37-4b3c-9c91-80ab4011f751

Problem Webpack, see how popup is open and closed, and then a new popup is open with the transaction.

https://github.com/user-attachments/assets/8165b3e7-2ed4-4704-91dc-82776edb0bd8

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
+  DAPP_URL,
+  defaultGanacheOptions,
   openDapp,
   unlockWallet,
-  DAPP_URL,
-  DAPP_ONE_URL,
-  regularDelayMs,
   WINDOW_TITLES,
-  defaultGanacheOptions,
-  switchToNotificationWindow,
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:11]: removed regular delay and reordered alphabetically

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 2]
```diff
@@ -51,17 +49,15 @@ describe('Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx', function () {
         await driver.findClickableElement({ text: 'Connect', tag: 'button' });
         await driver.clickElement('#connectButton');
 
-        await driver.delay(regularDelayMs);
-
-        await switchToNotificationWindow(driver);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:56]: removed deprecated method and added the new one (no need to add the delay)

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 3]
```diff
@@ -80,9 +76,6 @@ describe('Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx', function () {
           css: 'p',
         });
 
-        // Wait for the first dapp's connect confirmation to disappear
-        await driver.waitUntilXWindowHandles(2);
-
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:85]: no need to add this, as we handle the close window into the `clickElementAndWaitForWindowToClose

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 4]
```diff
@@ -125,28 +116,26 @@ describe('Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx', function () {
...
         await driver.clickElement({ text: 'Switch network', tag: 'button' });
 
         // Wait for switch confirmation to close then tx confirmation to show.
-        await driver.waitUntilXWindowHandles(3);
-        await driver.delay(regularDelayMs);
+        const newDialogWindow =
+          await driver.getNewNotificationHandleAfterOldOneCloses({
+            driver,
+            windowsBefore,
+          });
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:130]: here is where it happened the race condition, making the test fail. We now use the new driver function to mitigate this

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 5]
```diff
@@ -121,31 +112,39 @@ describe('Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx', function () {
...
-          tag: 'button',
-        });
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
 
         await driver.clickElement({ text: 'Switch network', tag: 'button' });
 
+        await driver.switchToWindowWithUrl(DAPP_ONE_URL);
+
         // Wait for switch confirmation to close then tx confirmation to show.
-        await driver.waitUntilXWindowHandles(3);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:137]: this point is where test breaks in webpack: if the next dialog window is opened really fast, this condition is never met

File: test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js [hunk 6]
```diff
@@ -276,31 +268,37 @@ describe('Request Queuing Dapp 1, Switch Tx -> Dapp 2 Send Tx', function () {
...
-          text: 'Cancel',
-          tag: 'button',
-        });
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
 
         await driver.clickElement({ text: 'Cancel', tag: 'button' });
+        await driver.switchToWindowWithUrl(DAPP_ONE_URL);
 
         // Wait for switch confirmation to close then tx confirmation to show.
-        await driver.waitUntilXWindowHandles(3);
```
Comment (seaona) [inline test/e2e/tests/request-queuing/dapp1-switch-dapp2-send.spec.js:292]: this point is where test breaks in webpack: if the next dialog window is opened really fast, this condition is never met

**Review by chloeYue — APPROVED** [review]
Looks good for me the fix ! Thanks !

**Review by DDDDDanica — APPROVED** [review]

**Pull request merged**